pkg/verifier/safe_delete.go

```diff
@@ -14,10 +15,14 @@ import (
 	"github.com/pkg/errors"
 )
 
-// SafeDelete moves block to backup bucket and if succeeded, removes it from source bucket.
-// It returns error if block dir already exists in backup bucket (blocks should be immutable) or any
-// of the operation fails.
-func SafeDelete(ctx context.Context, logger log.Logger, bkt objstore.Bucket, backupBkt objstore.Bucket, id ulid.ULID) error {
+// SafeDelete moves a TSDB block to a backup bucket and, on success, removes
+// it from the source bucket.  It returns error if block dir already exists in
+// the backup bucket (blocks should be immutable) or if any of the operations
+// fail.  When useExistingTempdir is true do not download the block but use the
+// previously downloaded block found in tempdir to avoid repeated downloads.
+// If tempdir is an empty string a unique temporary directory is created for
+// scratch space, otherwise the given directory is used.
+func SafeDelete(ctx context.Context, logger log.Logger, bkt objstore.Bucket, backupBkt objstore.Bucket, id ulid.ULID, useExistingTempdir bool, tempdir string) error {
```

Comment:
Couldn't we just get rid of `useExistingTempdir` here and use `tempdir` to control how this function works? It seems like we are making this unnecessarily complex. Perhaps I am missing something.

Reply:
I felt like giving a particular value of `tempdir` special meaning was not obvious enough. I think I prefer this over tempdir being empty. I do agree that below the if doesn't need the `tempdir == ""` case.

Reply:
If you feel strongly about this @GiedriusS I'm ok with the other behavior, just feel I will have forgotten the implicit behavior of the empty string in 3 months, hence I wanted it to be more explicit.

Reply:
Hm, I feel strongly about this - boolean in random place of the functions IMO smells bad. Same with 2 flags controlling same thing in very strict way.Essentially it does not make sense to set `useExistingTempdir = true` and `tempdir = ""` so we need to validate that etc. Interface is broader then needed.

Also if userExistingTempdir = true and tempdir is specified you literally don't use `bkt` so it's even worse in terms of readability and surprises ):

I would propose: 
*  `BackupAndDelete(ctx context.Context, logger log.Logger, bkt objstore.Bucket, backupBkt objstore.Bucket, id ulid.ULID)` as it was.
* `BackupAndDeleteFromLocal(ctx context.Context, logger log.Logger, bdir string, backupBkt objstore.Bucket)`

This will ensure separation of concerns AND in fact `BackupAndDelete` can use `BackupAndDeleteFromLocal` internally. Right now what I can see is 2 function baked in one function (:

Reply:
Yeah I like @bwplotka's suggestion best as well.